src/merkle/smt.rs

```diff
@@ -0,0 +1,187 @@
+///
+/// [SparseMerkleTree] currently doesn't support optimizations that compress Merkle proofs.
+pub trait SparseMerkleTree<const DEPTH: u8> {
+    /// The type for a key, which must be convertible into a `u64` infaillibly
+    type Key: Into<LeafIndex<DEPTH>> + Clone;
+    /// The type for a value
+    type Value: Clone + Default + PartialEq;
+    /// The type for a leaf
+    type Leaf;
+
+    // PROVIDED METHODS
+    // ---------------------------------------------------------------------------------------------
+
+    /// Returns a Merkle path from the leaf node specified by the key to the root.
+    ///
+    /// The node itself is not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if the specified index is too large given the depth of this Merkle tree.
+    fn get_merkle_path(&self, key: Self::Key) -> MerklePath {
+        let mut index: NodeIndex = {
+            let leaf_index: LeafIndex<DEPTH> = key.into();
+            leaf_index.into()
+        };
+
+        let mut path = Vec::with_capacity(index.depth() as usize);
+        for _ in 0..index.depth() {
+            let is_right = index.is_value_odd();
+            index.move_up();
+            let InnerNode { left, right } = self.get_inner_node(index);
+            let value = if is_right { left } else { right };
+            path.push(value);
+        }
+
+        MerklePath::new(path)
+    }
+
+    /// Updates value of the leaf at the specified index returning the old leaf value.
+    ///
+    /// This also recomputes all hashes between the leaf and the root, updating the root itself.
+    fn update_leaf_at(&mut self, key: Self::Key, value: Self::Value) -> Self::Value {
+        let old_value = self.insert_leaf_node(key.clone(), value.clone()).unwrap_or_default();
+
+        // if the old value and new value are the same, there is nothing to update
+        if value == old_value {
+            return value;
+        }
+
+        let leaf = self.get_leaf_at(&key);
+        let node_index = {
+            let leaf_index: LeafIndex<DEPTH> = key.into();
+            leaf_index.into()
+        };
+
+        self.recompute_nodes_from_index_to_root(node_index, Self::hash_leaf(&leaf));
+
+        old_value
+    }
+
+    /// Recomputes the branch nodes (including the root) from `index` all the way to the root.
+    /// `node_hash_at_index` is the hash of the node stored at index.
+    fn recompute_nodes_from_index_to_root(
+        &mut self,
+        mut index: NodeIndex,
+        node_hash_at_index: RpoDigest,
+    ) {
+        let mut value = node_hash_at_index;
+        for _ in 0..index.depth() {
+            let is_right = index.is_value_odd();
+            index.move_up();
+            let InnerNode { left, right } = self.get_inner_node(index);
+            let (left, right) = if is_right { (left, value) } else { (value, right) };
+            self.insert_inner_node(index, InnerNode { left, right });
+            value = Rpo256::merge(&[left, right]);
```

Comment:
One potential thing to change here is to first compute `value` and if the value is equal to an empty root at the current depth, remove the node from the tree rather than add it.

The main idea here is that it doesn't make sense to add empty roots as we assume that any node that is not present is empty.

Reply:
Created #246